File: sdks/kotlin/core/src/main/kotlin/io/getstream/visionagents/core/Router.kt
package io.getstream.visionagents.core

import io.getstream.visionagents.core.generated.SearchAnswer as AnswerSchema
import io.getstream.visionagents.core.generated.SearchDepth
import io.getstream.visionagents.core.generated.SearchOptions as OptionsSchema
import io.getstream.visionagents.core.generated.SearchRequest
import io.getstream.visionagents.core.generated.SearchResult

/**
 * Looking something up, configured once.
 *
 * Search is the one routed modality a device may reach: a question and its answer are one
 * round trip, and the answer is for whoever asked rather than for the app.
 *
 *     val router = agents.router(config = "healthcare")
 *     val answer = router.search("what changed in the pricing page")
 *
 * Transcription, a voice and a model are not here. Those run over the per-modality socket,
 * which the router refuses a device, so a pipeline of your own belongs to a backend.
 */
public class Router(
    public val backend: Backend,
    /** A stored router config, by name or id. Empty means every call says what it wants. */
    public val config: String = "",
    /** Cost labels carried onto everything routed here, on top of the config's own. */
    public val tags: Map<String, String> = emptyMap(),
) {
    /** Answers one question out of what is true now. */
    public suspend fun search(query: String, options: SearchOptions = SearchOptions()): SearchAnswer {
        val request = SearchRequest(
            query = query,
            configId = config.ifEmpty { null },
            options = options.schema,
            tags = tags.ifEmpty { null },
        )
        val answer = backend.post(
            listOf("v1", "search"),
            SearchRequest.serializer(),
            request,
            AnswerSchema.serializer(),
        )
        return SearchAnswer.of(answer)
    }
}

/**
 * How to find out today's answers.
 *
 * Null means the named router config decides, and the router decides what the config does not.
 * A provider that cannot express a term refuses the request rather than dropping it.
 */
public data class SearchOptions(
    /** A provider/model or a capability shortcut such as `search-fast`. */
    val target: String? = null,
    /**
     * How much work a search is worth: [Depth.Instant] answers from the index in a few hundred
     * milliseconds, [Depth.Deep] reads what it finds and can take tens of seconds.
     */
    val depth: Depth? = null,
    /** How many hits to return. */
    val results: Int? = null,
    /** Only answer from these domains. */
    val includeDomains: List<String> = emptyList(),
    val excludeDomains: List<String> = emptyList(),
    /** The kind of source to prefer, for the providers that classify their index. */
    val category: String? = null,
    /** How stale a cached page may be. Zero forces a live crawl. */
    val maxAgeHours: Int? = null,
    /** Country or region to answer from. */
    val location: String? = null,
) {
    public enum class Depth { Instant, Fast, Standard, Deep }

    internal val schema: OptionsSchema
        get() = OptionsSchema(
            target = target,
            depth = depth?.let { SearchDepth.valueOf(it.name.lowercase()) },
            results = results,
            includeDomains = includeDomains.ifEmpty { null },
            excludeDomains = excludeDomains.ifEmpty { null },
            category = category,
            maxAgeHours = maxAgeHours,
            location = location,
        )
}

/** What a search found. */
public data class SearchAnswer(
    val provider: String,
    val model: String,
    /** The provider's own summary, where it offers one: a sentence to say rather than a page to read. */
    val answer: String,
    /** The sources behind it, most relevant first. */
    val results: List<SearchHit>,
) {
    internal companion object {
        fun of(schema: AnswerSchema): SearchAnswer = SearchAnswer(
            provider = schema.provider,
            model = schema.model,
            answer = schema.answer.orEmpty(),
            results = schema.results.map(SearchHit::of),
        )
    }
}

/** One source a search found. */
public data class SearchHit(
    val title: String,
    val url: String,
    /** The relevant extract, which is what a model reads. */
    val text: String,
    val score: Double,
) {
    internal companion object {
        fun of(schema: SearchResult): SearchHit = SearchHit(
            title = schema.title.orEmpty(),
            url = schema.url,
            text = schema.text.orEmpty(),
            score = schema.score?.toDouble() ?: 0.0,
        )
    }
}

File: sdks/dart/core/lib/src/router.dart
import 'generated/api.dart' as api;

/// How much work a search is worth.
enum SearchDepth {
  /// From the index, in a few hundred milliseconds.
  instant,
  fast,
  standard,

  /// Reads what it finds, and can take tens of seconds.
  deep,
}

/// How to find out today's answers.
///
/// Everything is optional because everything has an answer already: a named router config
/// decides what this does not say, and the router decides what the config does not. A
/// provider that cannot express a term refuses the request rather than dropping it.
final class SearchOptions {
  const SearchOptions({
    this.target,
    this.providers = const [],
    this.depth,
    this.results,
    this.includeDomains = const [],
    this.excludeDomains = const [],
    this.category,
    this.maxAgeHours,
    this.location,
  });

  /// A provider/model or a capability shortcut such as `search-fast`.
  final String? target;

  /// Where to try, in order, which wins over [target] and [depth].
  final List<String> providers;
  final SearchDepth? depth;

  /// How many hits to return.
  final int? results;

  /// Only answer from these domains.
  final List<String> includeDomains;
  final List<String> excludeDomains;

  /// The kind of source to prefer, for the providers that classify their index.
  final String? category;

  /// How stale a cached page may be. Zero forces a live crawl.
  final int? maxAgeHours;

  /// Country or region to answer from.
  final String? location;

  api.SearchOptions? get _wire {
    final wire = api.SearchOptions(
      target: target,
      providers: providers.isEmpty ? null : providers,
      depth: depth?.name,
      results: results,
      includeDomains: includeDomains.isEmpty ? null : includeDomains,
      excludeDomains: excludeDomains.isEmpty ? null : excludeDomains,
      category: category,
      maxAgeHours: maxAgeHours,
      location: location,
    );
    return wire.toJson().isEmpty ? null : wire;
  }
}

/// What a search found.
final class SearchAnswer {
  const SearchAnswer({
    required this.provider,
    required this.model,
    this.answer = '',
    this.results = const [],
  });

  final String provider;
  final String model;

  /// The provider's own summary, where it offers one: a sentence to say rather than a page.
  final String answer;

  /// The sources behind it, most relevant first.
  final List<SearchHit> results;
}

/// One source a search found.
final class SearchHit {
  const SearchHit({required this.url, this.title = '', this.text = '', this.score = 0});

  final String url;
  final String title;

  /// The relevant extract, which is what a model reads.
  final String text;
  final double score;
}

/// Looking something up, configured once.
///
/// Search is the one routed modality a device may reach: a question and its answer are one
/// round trip, and the answer is for whoever asked. Transcription, a voice and a model on
/// their own run over a socket the router refuses to a device.
final class SearchRouter {
  SearchRouter(this._operations, {this.config = '', this.tags = const {}});

  final api.Operations _operations;

  /// A stored router config, by name or id. Without one, every call says what it wants.
  final String config;

  /// Cost labels carried onto everything routed here, on top of the config's own.
  final Map<String, String> tags;

  /// Answers one question out of what is true now.
  Future<SearchAnswer> search(String query, [SearchOptions options = const SearchOptions()]) async {
    final answer = await _operations.search(
      body: api.SearchRequest(
        query: query,
        configId: config.isEmpty ? null : config,
        options: options._wire,
        tags: tags.isEmpty ? null : tags,
      ),
    );
    return SearchAnswer(
      provider: answer.provider,
      model: answer.model,
      answer: answer.answer ?? '',
      results: [
        for (final hit in answer.results)
          SearchHit(
            url: hit.url,
            title: hit.title ?? '',
            text: hit.text ?? '',
            score: hit.score ?? 0,
          ),
      ],
    );
  }
}
